### PR TITLE
updated attendance, employee, recruitment

### PR DESCRIPTION
**Pull request by Ruhan-Ahmad**

### CODE REVIEW

File: assets/serializers.py [hunk 1]
```diff
@@ -15,6 +15,15 @@ class Meta:
         model = Asset
         fields = ["url", "assignee", "id", "title", "description", "asset_model", "asset_type", "cost"]
 
+    def to_representation(self, instance):
+        ret = super(AssetSerializer, self).to_representation(instance)
```
Comment (TheUsmanMirza) [inline assets/serializers.py:19]: instance.assignee.all().get() -> instance.assignee.get() they are the same.

File: attendance/models.py [hunk 2]
```diff
@@ -25,12 +25,26 @@ def __str__(self):
...
+        ('MARRIAGE_LEAVE', 'MARRIAGE_LEAVE'),
+    ]
     employee = models.ForeignKey(Employee, on_delete=models.CASCADE, related_name="leaves")
-    leave_type = models.CharField(max_length=50)
+    leave_type = models.CharField(max_length=50, choices=LEAVE_CHOICES)
     reason = models.TextField(max_length=500)
     request_date = models.DateTimeField()
-    from_date = models.DateField()
-    to_date = models.DateField()
+    from_date = models.CharField(max_length=50)
```
Comment (TheUsmanMirza) [inline attendance/models.py:45]: Charfield instead of DateFiled?

File: attendance/serializers.py [hunk 3]
```diff
@@ -1,17 +1,71 @@
...
     class Meta:
         model = Attendance
         fields = ["id", "employee", "check_in", "check_out", "status"]
 
+    def to_representation(self, instance):
+        ret = super(AttendanceSerializer, self).to_representation(instance)
+        ret['employee_name'] = str(str(instance.employee.first_name).capitalize() + " " +
+                                   str(instance.employee.last_name).capitalize())
+        ret['time_check_in'] = str(str(instance.check_in.hour + 5).zfill(2) + ":" +
+                                   str(instance.check_in.minute).zfill(2) +
```
Comment (TheUsmanMirza) [inline attendance/serializers.py:18]: treat date type fields as DateTime Field, and user parsers for that if possible.

File: attendance/serializers.py [hunk 4]
```diff
@@ -1,17 +1,71 @@
...
+        date1 = datetime.strptime(from_date, '%Y-%m-%d')
+        date2 = datetime.strptime(to_date, '%Y-%m-%d')
+
+        delta = date2 - date1
+        return delta.days
+
+    def to_representation(self, instance):
+        ret = super(LeaveSerializer, self).to_representation(instance)
+
+        ret['employee_name'] = str(str(instance.employee.first_name).capitalize() + " " +
```
Comment (TheUsmanMirza) [inline attendance/serializers.py:59]: if there is need to name string that much, its better to make it a property in user model and format name there.

File: attendance/views.py [hunk 5]
```diff
@@ -58,15 +70,16 @@ def check_today_attendance(self, request):
         if attendance:
             serializer = AttendanceSerializer(attendance, many=True, context=serializer_context)
             return Response(serializer.data[0], status=status.HTTP_200_OK)
-        return JsonResponse({'detail': 'You did not check-in today'}, status=status.HTTP_404_NOT_FOUND)
+        return JsonResponse({'detail': 'You did not check-in today'}, status=status.HTTP_204_NO_CONTENT)
```
Comment (TheUsmanMirza) [inline attendance/views.py:73]: 404 is the correct error

File: employees/models.py [hunk 6]
```diff
@@ -35,8 +35,8 @@ class Employee(User):
     designation = models.CharField(max_length=50)
     bank = models.CharField(max_length=50)
     account_number = models.TextField(max_length=50)
-    profile_pic = models.URLField(max_length=200)
-    joining_date = models.DateTimeField()
+    profile_pic = models.CharField(max_length=200)
+    joining_date = models.CharField(max_length=50)
```
Comment (TheUsmanMirza) [inline employees/models.py:39]: again datetime field

File: finance/serializers.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+class PayRollSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Payroll
+        fields = ["id", "basic_salary", "bonus", "reimbursement", "travel_allowance", "tax_deduction", "month", "year",
+                  "released", "employee"]
+
+    def to_representation(self, instance):
+        ret = super(PayRollSerializer, self).to_representation(instance)
+        ret['employee_name'] = str(str(instance.employee.first_name).capitalize() + " " +
```
Comment (TheUsmanMirza) [inline finance/serializers.py:14]: user property name

File: finance/views.py [hunk 8]
```diff
@@ -1,3 +1,58 @@
...
+            serializer = PayRollSerializer(payroll, many=True, context=serializer_context)
+            return Response(serializer.data, status=status.HTTP_200_OK)
+        return JsonResponse({'detail': 'No payroll is created for you yet'}, status=status.HTTP_204_NO_CONTENT)
+
+    def list(self, request, *args, **kwargs):
+        month = self.request.query_params.get('month')
+        year = self.request.query_params.get('year')
+        emp = self.request.query_params.get('emp')
+
+        if month or year or emp:
```
Comment (TheUsmanMirza) [inline finance/views.py:36]: use django filters instead.

**Review by TheUsmanMirza — CHANGES_REQUESTED** [review]
@Ruhan-Ahmad I have added few comments on the backend part. Plase give it a look
- and remove the commented/unused code.

File: attendance/views.py (hunk 5 above)
Comment (Ruhan-Ahmad) [reply attendance/views.py:73]: yes but if on employee login if employee first came to dashboard on new day it gives 404 error as he did not check in for today that's why I change it to 204 error.

File: employees/serializers.py [hunk 9]
```diff
@@ -25,14 +31,28 @@ class EmployeeSerializer(serializers.ModelSerializer):
...
+        ret['department'] = str(instance.department)
+        return ret
 
     def validate_password(self, value):
         if value:
+            email_body = 'Hi ' + self.initial_data['username'] + '!\n' + \
+                         ' Here is your password for HRMS Portal \n' + value
+            data = {'email_body': email_body, 'to_email': self.initial_data['email'],
+                    'email_subject': 'Password'}
+            generate_and_send_employee_credentials(data)
```
Comment (TheUsmanMirza) [inline employees/serializers.py:56]: make it a celery task and use .delay()

File: finance/serializers.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    def create(self, validated_data):
+        user_email = validated_data['employee'].email
+        email_body = 'Hi ' + validated_data['employee'].username + '!\n' + \
+                     ' Your payroll has been generated for the month of ' + validated_data['month']\
+                     + ' ' + validated_data['year'] + '\n' + \
+                     'You can now view it at your dashboard.'
+        data = {'email_body': email_body, 'to_email': user_email,
+                'email_subject': 'Payroll Generated'}
+        generate_and_send_employee_credentials(data)
```
Comment (TheUsmanMirza) [inline finance/serializers.py:25]: .delay()

File: finance/serializers.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+                    'email_subject': 'Payroll Updated'}
+        else:
+            email_body = 'Hi ' + validated_data['employee'].username + '!\n' + \
+                         ' Your payroll has been released for the month of ' + validated_data['month'] \
+                         + ' ' + validated_data['year'] + '\n' + \
+                         'You can now view it at your dashboard.'
+            data = {'email_body': email_body, 'to_email': user_email,
+                    'email_subject': 'Payroll Released'}
+
+        generate_and_send_employee_credentials(data)
```
Comment (TheUsmanMirza) [inline finance/serializers.py:48]: .delay()

File: user/views.py [hunk 12]
```diff
@@ -40,7 +42,12 @@ def post(self, request):
         data = {'email_body': email_body, 'to_email': user.email,
                 'email_subject': 'Verify your email'}
 
-        send_email.delay(data)
+        # send_email.delay(data)
+
+        # send email
```
Comment (TheUsmanMirza) [inline user/views.py:47]: use `generate_and_send_employee_credentials` method with .delay()

File: user/views.py [hunk 13]
```diff
@@ -90,17 +97,19 @@ def post(self, request):
...
+            email_body = 'Hi, \n Use link below to reset your password  \n' + \
+                         abs_url
             data = {'email_body': email_body, 'to_email': user.email,
-                    'email_subject': 'Reset your passsword'}
-            send_email.delay(data)
+                    'email_subject': 'Reset your password'}
+            # send_email.delay(data)
+
+            # send email
+            email = EmailMessage(
```
Comment (TheUsmanMirza) [inline user/views.py:110]: use `generate_and_send_employee_credentials` with .delay()

**Pull request merged**